chainstate/types/src/block_status.rs

```diff
@@ -13,48 +13,185 @@
     FullyChecked,
 }
 
 #[derive(Copy, Clone, PartialEq, Eq, Debug, Decode, Encode)]
-pub struct BlockStatus {
-    last_successful_validation_stage: BlockValidationStage,
-}
+pub struct BlockStatus(BlockStatusInternal);
 
 impl BlockStatus {
-    pub const FULLY_CHECKED: BlockStatus = Self::new_at_stage(BlockValidationStage::FullyChecked);
+    pub fn new_at_stage(stage: BlockValidationStage) -> Self {
+        let mut this = Self::new();
+        this.set_last_valid_stage(stage);
+        this
+    }
 
-    pub const fn new_at_stage(stage: BlockValidationStage) -> Self {
-        Self {
-            last_successful_validation_stage: stage,
-        }
+    pub fn new_fully_checked() -> Self {
+        Self::new_at_stage(BlockValidationStage::FullyChecked)
     }
 
-    pub const fn new() -> Self {
-        Self::new_at_stage(BlockValidationStage::Unchecked)
+    pub fn new() -> Self {
+        Self(BlockStatusInternal::new())
     }
 
     /// Advance the last successful validation stage to the specified value.
     /// Note that the stage can only be advanced one step at a time.
     pub fn advance_validation_stage_to(&mut self, new_stage: BlockValidationStage) {
-        assert!(self.last_successful_validation_stage.next() == Some(new_stage));
-        self.last_successful_validation_stage = new_stage;
+        assert!(self.last_valid_stage().next() == Some(new_stage));
+        self.set_last_valid_stage(new_stage);
     }
 
-    pub fn is_valid(&self) -> bool {
-        self.last_successful_validation_stage == BlockValidationStage::FullyChecked
+    pub fn last_valid_stage(&self) -> BlockValidationStage {
+        let val = self.0.get_field(BlockStatusField::ValidationStage);
+        BlockValidationStage::from_u64(val).expect("Corrupted BlockValidationStage")
     }
 
-    pub fn last_valid_stage(&self) -> BlockValidationStage {
-        self.last_successful_validation_stage
+    // Note: it's better to keep this function private if possible.
+    fn set_last_valid_stage(&mut self, stage: BlockValidationStage) {
+        self.0.set_field(
+            BlockStatusField::ValidationStage,
+            stage as BlockStatusEffectiveType,
+        );
+    }
+
+    pub fn is_fully_valid(&self) -> bool {
+        self.last_valid_stage() == BlockValidationStage::FullyChecked && self.is_ok()
+    }
+
+    pub fn is_ok(&self) -> bool {
+        !(self.validation_failed() || self.has_invalid_parent())
+    }
+
+    pub fn set_validation_failed(&mut self) {
+        self.0.set_flag(BlockStatusField::ValidationFailedBit, true);
+    }
+
+    pub fn validation_failed(&self) -> bool {
+        self.0.get_flag(BlockStatusField::ValidationFailedBit)
+    }
+
+    pub fn set_has_invalid_parent(&mut self) {
+        self.0.set_flag(BlockStatusField::InvalidParentBit, true);
+    }
+
+    pub fn has_invalid_parent(&self) -> bool {
+        self.0.get_flag(BlockStatusField::InvalidParentBit)
+    }
+
+    // Note: this is needed for testing only.
```

Comment:
Maybe you can add `#[cfg(test)]` on this? It may not work accross crates, but give it a shot.

Reply:
I've tried that, it doesn't work unfortunately.